### PR TITLE
Add support for bandlimited angular spectrum.

### PR DESCRIPTION
**Pull request by ebezzam**
@GJBoth rough example and implementation to add support for BLAS (bandlimited angular spectrum)

References:
- original paper: https://opg.optica.org/oe/fulltext.cfm?uri=oe-17-22-19662&id=186848#g002
- off-axis: https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150#articleTables

I've tried to create an example that replicates Fig. 9a of the [original paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-17-22-19662&id=186848#g002). It's not exactly the same (but close). We do see the high-frequency noise found in the original AS being suppressed.

propagation_comparison

TODO:
- [ ] implement direct/numerical integration as a ground-truth for comparison, as they do in Fig 4 and 6 of [original paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-17-22-19662&id=186848#g002). Example implementation [here](https://github.com/ebezzam/waveprop/blob/a2d65116336bfb6e95732fd982e5c3ec2109cff3/waveprop/rs.py#L33)
- [ ] verify off-axis

### CODE REVIEW

File: src/chromatix/functional/propagation.py [hunk 1]
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
...
+            v0 = (v_limit_p - v_limit_n) / 2
+            v_width = v_limit_p + v_limit_n
+
+        fx_max = u_width / 2
+        fy_max = v_width / 2
+
+        # bandlimit
+        H_filter = (np.abs(fX - u0) <= fx_max) * (np.abs(fY - v0) < fy_max)
+
+        # to jax
```
Comment (ebezzam) [inline src/chromatix/functional/propagation.py:322]: probably a better way to do above with Jax?

File: src/chromatix/functional/propagation.py [hunk 2]
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
...
+        u_limit_n = ((x0 - 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sx < x0:
+            u0 = (u_limit_p + u_limit_n) / 2
+            u_width = u_limit_p - u_limit_n
+        elif x0 <= -Sx:
+            u0 = -(u_limit_p + u_limit_n) / 2
+            u_width = u_limit_n - u_limit_p
+        else:
+            u0 = (u_limit_p - u_limit_n) / 2
+            u_width = u_limit_p + u_limit_n
```
Comment (GJBoth) [inline src/chromatix/functional/propagation.py:301]: We have to be careful here - jax _really_ doesn't like conditionals based on values, so we shouldn't do this. Is there some way to write this without if/else?

File: src/chromatix/functional/propagation.py [hunk 3]
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
...
+        v_limit_n = ((y0 - 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sy < y0:
+            v0 = (v_limit_p + v_limit_n) / 2
+            v_width = v_limit_p - v_limit_n
+        elif y0 <= -Sy:
+            v0 = -(v_limit_p + v_limit_n) / 2
+            v_width = v_limit_n - v_limit_p
+        else:
+            v0 = (v_limit_p - v_limit_n) / 2
+            v_width = v_limit_p + v_limit_n
```
Comment (GJBoth) [inline src/chromatix/functional/propagation.py:314]: Would there be some way to do for this x and y at the same time by vectorizing it?

File: src/chromatix/functional/propagation.py (hunk 1 above)
Comment (diptodip) [reply src/chromatix/functional/propagation.py:322]: This is making a rectangle mask in the frequency grid? If so we shouldn't need `jnp.where` and this should just work because it's a binary mask and not variably sized (assuming the lines above can have their conditionals removed).

File: src/chromatix/functional/propagation.py (hunk 2 above)
Comment (diptodip) [reply src/chromatix/functional/propagation.py:301]: Maybe using `jnp.sign(x0 - Sx)` or something like that?

**Comment by ebezzam** [comment]
@GJBoth and @diptodip thanks for your comments! I'll take a look.

One dummy question: is [off-axis](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150) really the same as your [kykx](https://github.com/chromatix-team/chromatix/blob/7304cd312b28eebc2f15c3c466e53074141d553b/src/chromatix/functional/propagation.py#L162)?

From what I understand `kykx` is tilting the plane wave illumination (which can shift the diffraction pattern like in the top row of Fig 7 of the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)), while off-axis is interested in sampling the destination field *off-axis* (Fig 1 and bottom row of Fig 7 in the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)).

File: src/chromatix/functional/propagation.py [hunk 4]
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
...
+        Sy, Sx = (1 / field.dk).squeeze()   # spatial dimension in microns
+        y0, x0 = (kykx / field.dk).squeeze()  # spatial shift in microns, TODO check
+        z0 = z.squeeze()  # propagation distance in microns
+        wv = field.spectrum.squeeze()  # wavelength in microns
+
+        dfX = 1.0 / Sx
+        dfY = 1.0 / Sy
+        N_y, N_x = field.spatial_shape
+        fX = np.linspace(-N_x // 2, N_x // 2 - 1, num=N_x)[np.newaxis, :] * dfX
+        fY = np.linspace(-N_y // 2, N_y // 2 - 1, num=N_y)[:, np.newaxis] * dfY
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply src/chromatix/functional/propagation.py]: done!

File: src/chromatix/functional/propagation.py (hunk 1 above)
Comment (ebezzam) [reply src/chromatix/functional/propagation.py]: Yes it's making a rectangular mask as shown in [Fig 5](https://opg.optica.org/getimagev2.cfm?img=7dKLEDezXUrkoUkzVOSYXrMOHiBaTdFKRs1SIJ54Pfo%3D&size=full&uri=oe-18-17-18453-g005) of the off-axis paper. I'll see if there's a way to remove the conditions before.

Comment (ebezzam) [reply src/chromatix/functional/propagation.py]: done!

File: src/chromatix/functional/propagation.py (hunk 2 above)
Comment (ebezzam) [reply src/chromatix/functional/propagation.py]: yep I think applying `jnp.sign` correctly does the trick

File: src/chromatix/functional/propagation.py (hunk 3 above)
Comment (ebezzam) [reply src/chromatix/functional/propagation.py]: done!

**Comment by ebezzam** [comment]
@GJBoth and @diptodip I think I've addressed your comments! It's now vectorized for y and x and without if-else conditions. I've left the original code commented because I wonder if it's easier for someone to read that than the current, very dense version?

Also I found a small bug in the previous version (I had an extra factor 2 because in the paper they talked about the field size being increased by 2x to linearize the convolution, but within your propagator function the field is already padded). So now the result looks more like the original paper.

propagation_comparison

Below is from original paper (maybe they apply a gamma correction?)

Screenshot 2024-05-25 at 6 02 01 PM

As mentioned above, I think off-axis may be different from your kykx for tilting the illumination? Can you confirm? If so, I think it could be interesting to add a `shift_yx` optional argument for shifting the destination sampling window as done in the off-axis paper.

File: examples/bandlimited_angular_spectrum.py
Comment (diptodip) [inline examples/bandlimited_angular_spectrum.py:1]: This is a great example but would you mind to move it to a notebook with the plots included under `docs/examples/bandlimited_angular_spectrum.ipynb`? I can do it myself so it's no big deal but that way people can see the plots and we can generate a page on the docs website for it.

File: src/chromatix/functional/propagation.py
Comment (diptodip) [inline src/chromatix/functional/propagation.py:1]: So like you said, shifting is different from just the `kykx` and like Rainer pointed out I think the best thing to do is to add a property to the `Field` to keep track of this. But we can do that later.

And I think it's better to get rid of the comments :)

I don't think the code is that dense and we can of course add some explanation comments of how the coordinates are being chosen.

And finally (not a big deal but just a note for myself later) we have `l1_norm` in the utilities that we use in `square_pupil` to make rectangular masks.

File: src/chromatix/field.py
Comment (diptodip) [inline src/chromatix/field.py:1]: Great add!

**Review by diptodip — CHANGES_REQUESTED** [review]

**Comment by diptodip** [comment]
@ebezzam This all looks awesome! Sorry I'm just now getting to see this. By the way, I'm going to update this PR to merge into `hackathon_2024` instead of `main`. I'll leave some more specific comments but in general:

> @GJBoth and @diptodip thanks for your comments! I'll take a look.
> 
> One dummy question: is [off-axis](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150) really the same as your [kykx](https://github.com/chromatix-team/chromatix/blob/7304cd312b28eebc2f15c3c466e53074141d553b/src/chromatix/functional/propagation.py#L162)?
> 
> From what I understand `kykx` is tilting the plane wave illumination (which can shift the diffraction pattern like in the top row of Fig 7 of the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)), while off-axis is interested in sampling the destination field _off-axis_ (Fig 1 and bottom row of Fig 7 in the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)).

Yes I think you're right about that. And Rainer also made the great point today that we could simply add a new property to our `Field` objects to keep track of a spatial location (destination sampling window  center as in the paper you shared). This way when we propagate off axis we can just update this property instead of requiring propagation of a large field of view (of course both could still be allowed).

So this I think is definitely a nice thing that we should add, but perhaps we work on that update in another PR?

> Below is from original paper (maybe they apply a gamma correction?)
> 
> Screenshot 2024-05-25 at 6 02 01 PM

It does look like a gamma correction. What does it look like if you plot your results with a gamma correction (e.g. `intensity**0.7` or so)? Or they plot amplitude and not intensity?

**Comment by ebezzam** [comment]
@diptodip thanks for your comments! I'll address them on Monday

**Comment by ebezzam** [comment]
@diptodip, @GJBoth I've addressed your comments, thanks for your support!

I've created a notebook which shows BLAS and off-axis propagation:

Screenshot 2024-05-28 at 3 02 41 PM


Screenshot 2024-05-28 at 3 03 02 PM

I've added a `shift_yx` optional parameter to `compute_asm_propagator` so that there is a working implementation of off-axis, but if you think it's better to move that to the `Field` class, it should be straightforward.

Let me know if you need anything else!

**Comment by diptodip** [comment]
Thanks Eric! We'll merge things into the hackathon_2024 branch and go from there. Really appreciate all your work implementing this!

**Review by diptodip — APPROVED** [review]

**Comment by RainerHeintzmann** [comment]
Cool stuff you added! Looks great. Also the off-axis. Nice.
